c10/cuda/CUDAException.cpp

```diff
@@ -1,35 +1,380 @@
 #include <c10/cuda/CUDAException.h>
 
+#include <c10/util/Backtrace.h>
 #include <c10/util/Exception.h>
+#include <c10/util/irange.h>
 #include <cuda_runtime.h>
 
+#include <algorithm>
+#include <memory>
+#include <stdexcept>
 #include <string>
+#include <thread>
+
+#define CHECK_CUDA_API_CALL_WITHOUT_CHECKING_DEVICE_ASSERTS \
+  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false)
 
 namespace c10 {
 namespace cuda {
 
+/// Check that kernels ran correctly by checking the message buffer. BLOCKING.
+std::string c10_retrieve_device_side_assertion_info() {
+#ifdef TORCH_USE_CUDA_DSA
+  const auto& launch_registry = CUDAKernelLaunchRegistry::get_singleton_ref();
+  if (!launch_registry.enabled) {
+    return "Device-side assertion tracking was not enabled by user.";
+  } else if (!launch_registry.do_all_devices_support_managed_memory) {
+    return "Device-side assertions disabled because not all devices support managed memory.";
+  }
+
+  // Hack that saves a lot of challenging sync logic.
+  // The GPU increments the number of errors it's observed and the CPU can see
+  // that happening immediately which means we can make it here before the GPU
+  // is done writing information about those errors to memory.
+  // A short pause gives it time to finish. Since something's gone wrong, this
+  // pause shouldn't affect perf.
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  // The snapshot causes a brief block. That's okay because this function only
+  // executes if something's gone wrong such that speed is no longer a priority.
+  const auto launch_data = launch_registry.snapshot();
+  const auto& assertion_data = launch_data.first;
+  const auto& launch_infos = launch_data.second;
+
+  std::stringstream oss;
+
+  oss << "# of GPUs this process interacted with = "
+      << launch_registry.gpus_interacted_with() << std::endl;
+
+  // Loop over each device that could be managed by the process
+  for (const auto device_num : c10::irange(assertion_data.size())) {
+    const auto& assertion_data_for_device = assertion_data.at(device_num);
+
+    // Did anything fail?
+    const auto failures_found = std::min(
+        assertion_data_for_device.assertion_count,
+        C10_CUDA_DSA_ASSERTION_COUNT);
+    if (failures_found == 0) {
+      continue;
+    }
+
+    // Something failed, let's talk about that
+    oss << failures_found
+        << " CUDA device-side assertion failures were found on GPU #"
+        << device_num << "!" << std::endl;
+    if (assertion_data_for_device.assertion_count >
+        C10_CUDA_DSA_ASSERTION_COUNT) {
+      oss << "But at least " << assertion_data_for_device.assertion_count
+          << " assertion failures occurred on the device" << std::endl;
+      oss << "Adjust `C10_CUDA_DSA_ASSERTION_COUNT` if you need more assertion failure info"
```

Comment:
Though kinds of messages often get out of date (as one edits the environment variable but ignores the rest

Reply:
Do you have a proposed solution?